Software/bsw/static/Mcal/CAN/src/Can.c

```diff
@@ -775,152 +775,261 @@ void Can_MainFunction_BusOff(void) {
+				/*
+				 * check if hardware is busy
+				 */
+				while (HWREG(ui32Base + CAN_O_IF1CRQ) & CAN_IF1CRQ_BUSY)
+				{}
+
+
+				ui16CmdMaskReg |= CAN_IF1CMSK_WRNRD | CAN_IF1CMSK_DATAA |
+								  CAN_IF1CMSK_DATAB | CAN_IF1CMSK_CONTROL|
+								  CAN_IF1CMSK_ARB ;
+
+				ui16MsgCtrl |= CAN_IF1MCTL_TXRQST;			/// transmission request
+				ui16ArbReg_2 = CAN_IF1ARB2_DIR;				/// transmit M.O.
+
+
+				if (g_Config_Ptr->CanConfigSetRef->CanHardwareTransmitObjectRef[hth_index].CanIdType == EXTENDED)
+				{
+
+					ui16ArbReg_1 |= PduInfo->id & CAN_IF1ARB1_ID_M ;
+					ui16ArbReg_2 |= (PduInfo->id >> 16 ) & CAN_IF1ARB2_ID_M ;
+					ui16ArbReg_2 |= CAN_IF1ARB2_MSGVAL | CAN_IF1ARB2_XTD ;
+
+				}
+				else if (g_Config_Ptr->CanConfigSetRef->CanHardwareTransmitObjectRef[hth_index].CanIdType == STANDARD)
+				{
+					ui16ArbReg_2 |= (  PduInfo->id  << 2 ) & CAN_IF1ARB2_ID_M ;
+					ui16ArbReg_1 |= CAN_IF1ARB2_MSGVAL ;
+
+				}
+				else
+				{}
+
+
+				ui16MsgCtrl |=  ( ( PduInfo->length ) & CAN_IF1MCTL_DLC_M ) ;
+
+				ui16MsgCtrl |= CAN_IF1MCTL_EOB  | CAN_IF1MCTL_TXIE ;
+
+
+
+				CANDataRegWrite (
+
+						PduInfo->sdu,
+						(uint32 *)( ui32Base + CAN_O_IF1DA1 ),
+						PduInfo->length
+				);
+
+
+				HWREG(ui32Base + CAN_O_IF1ARB2) &=~ CAN_IF1ARB2_MSGVAL;
+
+				HWREG(ui32Base + CAN_O_IF1CMSK) = ui16CmdMaskReg;
+				HWREG(ui32Base + CAN_O_IF1MSK1) = ui16MaskReg_1;
+				HWREG(ui32Base + CAN_O_IF1MSK2) = ui16MaskReg_2;
+				HWREG(ui32Base + CAN_O_IF1ARB1) = ui16ArbReg_1;
+				HWREG(ui32Base + CAN_O_IF1ARB2) = ui16ArbReg_2;
+				HWREG(ui32Base + CAN_O_IF1MCTL) = ui16MsgCtrl;
+
+				HWREG(ui32Base + CAN_O_IF1CRQ) = g_Config_Ptr->CanConfigSetRef->CanHardwareTransmitObjectRef[hth_index].CanObjectId ;
+
+				HTH_semaphore[hth_index] = 0 ;
 
 				/*
-				 *
-				 * call the Tivaware function now !
+				 * (SRS_Can_01049)
 				 */
-				CANMessageSet(uiBase, ui32ObjID, psMsgObject, eMsgType);
-				Saved_swPduHandle = PduInfo->swPduHandle; //// save the
-				returnVal = E_OK;
+				returnVal = E_OK ;
+			}
+			else
+			{
+				/*
+				 * [SWS_Can_00213] ⌈ The function Can_Write shall perform no actions if the
+				 * hardware transmit object is busy with another transmit request for an L-PDU:
+				 * 1. The transmission of the other L-PDU shall not be cancelled and the function
+				 * Can_Write is left without any actions.
+				 * 2. The function Can_Write shall return CAN_BUSY.⌋ (SRS_Can_01049).
+				 */
+				returnVal = CAN_BUSY ;
 
 			}
 
-		} else /////// CAN Is busy
+		}
+
+	}
+	return returnVal ;
+
+}
+
+STATIC void CANDataRegWrite ( uint8 * pui8Data, uint32 * pui32Register , uint8 ui8Size)
+{
+
+	uint8 index ;
+	uint32 ui32DataValue ;
+
+
+	for (index = 0 ; index < ui8Size;)
+	{
+
+		ui32DataValue = pui8Data[index++] ;
+
+		if (index < ui8Size)
 		{
-			returnVal = CAN_BUSY;
+			ui32DataValue |= ( pui8Data[index++] << 8 ) ;
 		}
+
+
+		HWREG(pui32Register++) = ui32DataValue;
+
 	}
 
-	return returnVal;
+
 }
 
+
+
+
+
```

Comment:
My question here, Is the team agreed of all this changes ? and why no one in the reviewer list ? :) 

Great work, keep it up, thanks for being the first one working in that :)

Reply:
yes, kind of :'D